### PR TITLE
Implement .entropy() methods for all distributions

### LINKED ISSUE #36: Implement differentiable .entropy() methods
- Normal, Bernoulli, Categorical, Exponential, MultivariateNormal are all easy to implement
- Gamma, Beta, Dirichlet all require `digamma` from https://github.com/pytorch/pytorch/pull/3955

Differentiable entropy is useful in variational inference.

### PR DESCRIPTION
**Pull request by fritzo**
Closes #36 

This also:
1. ~Fixes some bugs in torch's `TestCase.assertEqual()`~ (moved to @neerajprad 's PR)
2. Adds some infrastructure for writing generic tests in `test_distributions.py`

### CODE REVIEW

**Comment by fritzo** [comment]
CC @neerajprad could you please review the new generic testing tools?

**Comment by alicanb** [comment]
Shall we add it to the base class as well?

File: test/test_distributions.py [hunk 1]
```diff
@@ -15,6 +17,45 @@
     TEST_NUMPY = False
 
 
+# Register all distributions for generic tests.
+Example = namedtuple('Example', ['Dist', 'params'])
```
Comment (neerajprad) [inline test/test_distributions.py]: Looks good. Is the plan to use this mainly for shape testing? We may need to extend the fixture to support other tests (in later PRs).

Comment (fritzo) [reply test/test_distributions.py]: Yes, we can add more fields as we use this fixture for more tests (beyond just shape).

**Comment by fritzo** [comment]
> Shall we add it to the base class as well?

Done.  I've also implemented `.entropy()` for the three new distributions that merged today :smile:

File: test/test_distributions.py (hunk 1 above)
Comment (fritzo) [reply test/test_distributions.py]: BTW I have a slight ideological preference for this data-only approach to parameterized tests. I find it improves transparency to limit parameterized test data to plain old data, and keep the logic that manipulates that data in `TestCase` classes or other functions. I'd like to hear your opinions about test fixture style.

Comment (neerajprad) [reply test/test_distributions.py]: I don't really have too many opinions. I find it nice for fixtures to contain the bare minimum specification needed to run the tests - i.e. just pure data, and have test methods themselves or other helper utilities manipulate that further as the case may be. 

If there is a lot of setup needed to construct the fixture, then it could be abstracted into its own class or function - for instance constructing a JSON request for a service. If all methods need to do the same manipulation, it is again nice to abstract the fixture setup into a separate class or function. If each test manipulates the data somewhat differently, then the test methods itself should consume the raw fixture data. Once these fixtures start containing multiple fields (expected test values, etc.) it may get unwieldy and it may make sense to have separate fixtures for different tests rather than have a mega fixture.

**Comment by neerajprad** [comment]
Can this be merged now?

**Comment by fritzo** [comment]
I'm simply waiting for somebody's LGTM :smile: then I'll move the PR to pytorch/pytorch.

**Comment by fritzo** [comment]
Does anyone want to review the math before I move this PR to pytorch/pytorch? I've simply implemented `.entropy()` methods from wikipedia.

**Comment by neerajprad** [comment]
> I'm simply waiting for somebody's LGTM

My bad..I didn't realize that this wasn't already on PyTorch. I'll just briefly go through the `.entropy` calculations. Looks good to me otherwise.

File: torch/distributions/bernoulli.py [hunk 2]
```diff
@@ -46,6 +46,12 @@ def log_prob(self, value):
         # evaluate using the values
         return log_pmf.gather(-1, value.unsqueeze(-1).long()).squeeze(-1)
 
+    def entropy(self):
+        p = torch.stack([self.probs, 1.0 - self.probs])
+        p_log_p = torch.log(p) * p
+        p_log_p[p == 0] = 0
+        return -p_log_p.sum(0)
```
Comment (neerajprad) [inline torch/distributions/bernoulli.py]: Shouldn't this be summed along the last axis?

Comment (fritzo) [reply torch/distributions/bernoulli.py]: This sums out over the new dimension created by `torch.stack()` which is dimension 0. The default order for `torch.satck()` and is cheaper than any other dimension since the stacked tensors can be copied contiguously (rather than interleaved).

Comment (neerajprad) [reply torch/distributions/bernoulli.py]: Ahh..missed that. Makes sense.

**Comment by fritzo** [comment]
Moving this PR upstream to https://github.com/pytorch/pytorch/pull/4268
